lib/settings.coffee

```diff
@@ -0,0 +1,202 @@
+#
+# * Sync.set() and Sync.clear() propagate local changes to chrome.storage.sync.
+# * Sync.handleStorageUpdate() listens for changes to chrome.storage.sync and propagates those
+#   changes to localStorage and into vimium's internal state.
+# * Sync.fetchAsync() polls chrome.storage.sync at startup, similarly propagating
+#   changes to localStorage and into vimium's internal state.
+#
+# The effect is best-effort synchronization of vimium options/settings between
+# chrome/vimium instances.
+#
+# NOTE:
+#   Values handled within this module are ALWAYS already JSON.stringifed, so
+#   they're always non-empty strings.
+#
+
+root = exports ? window
+Sync =
+
+  storage: chrome.storage.sync
+  doNotSync: ["settingsVersion", "previousVersion"]
+
+  # This is called in main.coffee.
+  init: ->
+    chrome.storage.onChanged.addListener (changes, area) -> Sync.handleStorageUpdate changes, area
+    @fetchAsync()
+
+  # Asynchronous fetch from synced storage, called only at startup.
+  fetchAsync: ->
+    @storage.get null, (items) =>
+      unless chrome.runtime.lastError
+        for own key, value of items
+          Settings.storeAndPropagate key, value if @shouldSyncKey key
+
+  # Asynchronous message from synced storage.
+  handleStorageUpdate: (changes, area) ->
+    for own key, change of changes
+      Settings.storeAndPropagate key, change?.newValue if @shouldSyncKey key
+
+  # Only called synchronously from within vimium, never on a callback.
+  # No need to propagate updates to the rest of vimium, that's already been done.
+  set: (key, value) ->
+    if @shouldSyncKey key
+      setting = {}; setting[key] = value
+      @storage.set setting
+
+  # Only called synchronously from within vimium, never on a callback.
+  clear: (key) ->
+    @storage.remove key if @shouldSyncKey key
+
+  # Should we synchronize this key?
+  shouldSyncKey: (key) -> key not in @doNotSync
+
+#
+# Used by all parts of Vimium to manipulate localStorage.
+#
+
+# Select the object to use as the cache for settings.
+if Utils.isExtensionPage()
+  if Utils.isBackgroundPage()
+    settingsCache = localStorage
+  else
+    settingsCache = extend {}, localStorage # Make a copy of the cached settings from localStorage
+else
+  settingsCache = {}
+
+root.Settings = Settings =
+  cache: settingsCache
+  init: -> Sync.init()
```

Comment:
No. Starting to use `chrome.storage` (via `Sync`) everywhere *is the point* of this PR*, so we can stop moving settings data between contexts manually.